core/Enclave/key_operation.h

```diff
@@ -0,0 +1,92 @@
+ *      distribution.
+ *   3. Neither the name of Intel Corporation nor the names of its
+ *      contributors may be used to endorse or promote products derived
+ *      from this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+ */
+#include "log_utils.h"
+#include "sgx_tseal.h"
+
+#include <string>
+#include <stdio.h>
+#include <stdbool.h>
+#include <mbusafecrt.h>
+
+#include "sgx_report.h"
+#include "sgx_utils.h"
+#include "sgx_tkey_exchange.h"
+
+#ifndef _KEY_OPERATION_H_
+#define _KEY_OPERATION_H_
+
+void printf(const char *fmt, ...);
+
+sgx_status_t ehsm_aes_gcm_encrypt(const uint8_t *aad, size_t aad_len,
+                                  const uint8_t *cmk_blob, size_t SIZE_OF_KEYBLOB_T,
+                                  const uint8_t *plaintext, size_t plaintext_len,
+                                  uint8_t *cipherblob, size_t cipherblob_len,
+                                  ehsm_keyspec_t keyspec);
+
+sgx_status_t ehsm_aes_gcm_decrypt(const uint8_t *aad, size_t aad_len,
+                                  const uint8_t *cmk_blob, size_t SIZE_OF_KEYBLOB_T,
+                                  const uint8_t *cipherblob, size_t cipherblob_len,
+                                  uint8_t *plaintext, size_t plaintext_len,
+                                  ehsm_keyspec_t keyspec);
+
+sgx_status_t ehsm_sm4_encrypt(const uint8_t *aad, size_t aad_len,
+                              const uint8_t *cmk_blob, size_t SIZE_OF_KEYBLOB_T,
+                              const uint8_t *plaintext, size_t plaintext_len,
+                              uint8_t *cipherblob, size_t cipherblob_len,
+                              ehsm_keyspec_t keyspec);
+
+sgx_status_t ehsm_sm4_decrypt(const uint8_t *aad, size_t aad_len,
+                              const uint8_t *cmk_blob, size_t SIZE_OF_KEYBLOB_T,
+                              const uint8_t *cipherblob, size_t cipherblob_len,
+                              uint8_t *plaintext, size_t plaintext_len,
+                              ehsm_keyspec_t keyspec);
+
+sgx_status_t ehsm_asymmetric_encrypt(const ehsm_keyblob_t *cmk, ehsm_data_t *plaintext, ehsm_data_t *ciphertext);
```

Comment:
why you renmae it as ehsm_asymmetric_encrypt

Reply:
The plan is to merge rsa and sm2 ec/decryption into one function.
EVP should be able to do this, so I don't need to use encapsulated function